Add MapGeom.__eq__ operator

registerrier opened the pull request:
This PR introduces the `__eq__` operator on both `WcsGeom` and `HpxGeom`. 

`MapAxis` equality is tested as well as geometry consistency. In the case of `WcsGeom` we use a a tolerance of 1e-6 for WCS comparison. The latter is obtained with `~astrop.wcs.WCS.compare`.

A later PR will add the equality check for map arithmetics.

cdeil commented on gammapy/maps/hpx.py:
```diff
@@ -1732,6 +1732,31 @@ def __repr__(self):
+        if self.data_shape != other.data_shape:
+            return False
+
+        result = True
```
This is a little simpler, no?
```
for ax1, ax2 in zip(self.axes, other.axes):
    if ax1 != ax2:
        return False
```
Alternatively if you want to create a local status variable, I think this would be equivalent:
```
axes_equal = all(ax1 == ax2 for ax1, ax2 in zip(self.axes, other.axes))
```

registerrier replied:
OK. Changed to first solution

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -882,6 +882,20 @@ def __repr__(self):
+    def __eq__(self, other):
+        """Test equality between two `~gammapy.maps.WcsGeom`. Espect tolerance of 1e-6 for `~astropy.wcs.WCS` objects"""
+        if not isinstance(other, self.__class__):
+            raise TypeError('Cannot compare WcsGeom with {}'.format(other.__class__))
```
I think you're supposed to `return NotImplemented` when implementing `__eq__`, no?
Can you please check the errors you get and check the Python docs what's recommended?

registerrier replied:
OK. Done

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -882,6 +882,20 @@ def __repr__(self):
         return str_
 
+    def __eq__(self, other):
+        """Test equality between two `~gammapy.maps.WcsGeom`. Espect tolerance of 1e-6 for `~astropy.wcs.WCS` objects"""
```
Comment line is too long. IMO "Test equality between two `~gammapy.maps.WcsGeom`." doesn't add value and could be removed. The commend about WCS and 1e6 I would put inline above the line where you call the corresponding method.

registerrier replied:
Done

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -882,6 +882,20 @@ def __repr__(self):
+        if self.data_shape != other.data_shape:
+            return False
+
+        result = True
```
Also here: I would not introduce a local state variable "result" that you accumulate. Just check things and return False if not equal one by one, and put `return True` on the last line.
I think I've seen that coding pattern a lot in Python codes.

registerrier replied:
Done

cdeil commented on gammapy/maps/tests/test_wcs.py:
```diff
@@ -296,3 +296,36 @@ def test_get_axis_index_by_name():
     with pytest.raises(ValueError):
         geom.get_axis_index_by_name("time")
+
+test_axis1 = [MapAxis(nodes=(1,2,3,4), unit='TeV', node_type='center')]
```
Also here, I would suggest to try and reduce the lines of test code if possible.
We already have many 1000s too much and it's hard to maintain.

Concretely, I think we shouldn't even try to exhaustively test astropy.wcs.WCS properties, but just rely that that compare works, and just have 1 or 2 explicit cases where it's the same and we get True and one where it's off by 1e-3 and False.
And the whole `@pytest.mark.parametrize` can be removed and the test cases in the test function, if there's only one line for each the test parametrisation doesn't by you anything, no?

Matter of taste - feel free to finish up as you like an merge.

registerrier commented:
Implemented some suggested modifications.
Removed the exception raising and replaced it with `NotImplemented`. The associated tests were removed as well.
@cdeil Merge?
